controllers/datadogagent/feature/externalmetrics/feature.go

```diff
@@ -0,0 +1,337 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package externalmetrics
+
+import (
+	"fmt"
+	"strconv"
+
+	apicommon "github.com/DataDog/datadog-operator/apis/datadoghq/common"
+	apicommonv1 "github.com/DataDog/datadog-operator/apis/datadoghq/common/v1"
+	"github.com/DataDog/datadog-operator/apis/datadoghq/v1alpha1"
+	"github.com/DataDog/datadog-operator/apis/datadoghq/v2alpha1"
+	apiutils "github.com/DataDog/datadog-operator/apis/utils"
+	"github.com/DataDog/datadog-operator/controllers/datadogagent/component"
+	componentdca "github.com/DataDog/datadog-operator/controllers/datadogagent/component/clusteragent"
+	"github.com/DataDog/datadog-operator/controllers/datadogagent/feature"
+	"github.com/DataDog/datadog-operator/pkg/kubernetes/rbac"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	apiregistrationv1 "k8s.io/kube-aggregator/pkg/apis/apiregistration/v1"
+)
+
+func init() {
+	err := feature.Register(feature.ExternalMetricsIDType, buildExternalMetricsFeature)
+	if err != nil {
+		panic(err)
+	}
+}
+
+func buildExternalMetricsFeature(options *feature.Options) feature.Feature {
+	externalMetricsFeat := &externalMetricsFeature{}
+
+	return externalMetricsFeat
+}
+
+type externalMetricsFeature struct {
+	wpaController      bool
+	useDDM             bool
+	port               int32
+	url                string
+	keySecret          map[string]secret
+	serviceAccountName string
+	owner              metav1.Object
+}
+
+type secret struct {
+	name string
+	key  string
+	data []byte
+}
+
+// ID returns the ID of the Feature
+func (f *externalMetricsFeature) ID() feature.IDType {
+	return feature.ExternalMetricsIDType
+}
+
+// Configure is used to configure the feature from a v2alpha1.DatadogAgent instance.
+func (f *externalMetricsFeature) Configure(dda *v2alpha1.DatadogAgent) (reqComp feature.RequiredComponents) {
+	f.owner = dda
+	em := dda.Spec.Features.ExternalMetricsServer
+
+	if em != nil && apiutils.BoolValue(em.Enabled) {
+		f.wpaController = apiutils.BoolValue(em.WPAController)
+		f.useDDM = apiutils.BoolValue(em.UseDatadogMetrics)
+		f.port = *em.Port
+		if em.Endpoint != nil {
+			if em.Endpoint.URL != nil {
+				f.url = *em.Endpoint.URL
+			}
+			creds := em.Endpoint.Credentials
+			if creds != nil {
+				f.keySecret = make(map[string]secret)
+				if !v2alpha1.CheckAPIKeySufficiency(creds, apicommon.DDExternalMetricsProviderAPIKey) ||
+					!v2alpha1.CheckAppKeySufficiency(creds, apicommon.DDExternalMetricsProviderAppKey) {
```

Comment:
based on the comment below, should it be `&&` ?

Reply:
(or perhaps the map is meant to be pre-populated and overwritten in the lines below)

Reply:
Sorry, that comment was a bit misleading. The intent of that `if` statement is to check if either of `CheckXXXKeySufficiency` fails. This way, if, for some reason, someone decides to store their APP key in a secret but hardcode the API key (example below), we'll still create a secret for the API key. I'll try to make the comment clearer.

```yaml
    externalMetricsServer:
      enabled: true
      endpoint:
        credentials:
          apiKey: "<hardcoded-key>"
          appSecret:
            keyName: app-key
            secretName: super-secret
```

Reply:
ok thanks. i'm trying to think if there's a way to prevent calling each `v2alpha1.Check[...]KeySufficiency` twice. would it work to move lines 94-100 and 104-109 within this `if` block?

Reply:
I don't think I can move them since that would mean the secret's name+key won't be stored if both app key and api key have secrets defined for them. I think the `isSet` should be enough to check if there is a secret though so I can remove the `CheckAPI/AppKeySufficiency` calls